src/bpfilter/xlate/cli.c

```diff
@@ -56,6 +60,206 @@ int _bf_cli_ruleset_flush(const struct bf_request *request,
+             * input rule specified "COUNTERS" or not. Thus, advancing map_index
+             * is unconditional. When rule->counters == false, the corresponding
+             * counter value will remain {0,0}. */
+            counter_index++;
+            map_idx++;
+        }
+
+        // Chain's policy counter
+        r = bf_cgen_get_counter(cgen, BF_COUNTER_POLICY,
+                                &counters[counter_index]);
+        if (r < 0)
+            return bf_err_r(r, "Failed to get policy counter\n");
+        counter_index++;
+
+        // Chain's error counter
+        r = bf_cgen_get_counter(cgen, BF_COUNTER_ERRORS,
+                                &counters[counter_index]);
+        if (r < 0)
+            return bf_err_r(r, "Failed to get error counter\n");
+    }
+
+    return 0;
+}
+
+static int _bf_cli_get_chain_list(bf_list **chain_list)
+{
+    bf_list_ops ops = {// chain_list will only contain the chains, not own them.
+                       .free = (bf_list_ops_free)bf_nop_free,
+                       .marsh = (bf_list_ops_marsh)bf_chain_marsh};
+
+    int r = bf_list_new(chain_list, &ops);
+    if (r < 0)
+        return bf_err_r(r, "Failed to create the chain list");
+
+    // Loop over codegens which correspond to hooks
+    for (int i = 0; i < _BF_HOOK_MAX; ++i) {
+        struct bf_cgen *cgen;
+
+        cgen = bf_ctx_get_cgen(i, NULL);
+
+        if (!cgen || !cgen->chain)
+            continue;
+
+        r = bf_list_add_tail(*chain_list, cgen->chain);
+        if (r)
+            return bf_err_r(r, "Failed to add chain to list");
+    }
+
+    return 0;
+}
+
+static int _bf_cli_get_counters_marsh(struct bf_marsh **counter_marsh,
+                                      bf_list *chain_list, bool with_counters)
+{
+    int r;
+    size_t num_counters = 0;
+    struct bf_counter *counters = NULL;
+
+    bf_assert(bf_list_size(chain_list) > 0);
+
+    // If we don't want counters, return an empty marsh
+    if (!with_counters) {
+        r = bf_marsh_new(counter_marsh, NULL, 0);
+        return r;
+    }
+
+    /* Each chain has a policy counter and an error counter.
+     * Each rule has a counter, though it may be unused. */
+    num_counters =
+        (2 * bf_list_size(chain_list)) + _bf_cli_num_rules(chain_list);
+
+    // We must have at least have the two chain counters
+    bf_assert(num_counters >= 2);
+
+    counters = calloc(num_counters, sizeof(struct bf_counter));
+    if (!counters)
+        bf_err_r(-ENOMEM, "Failed to allocate memory for counters\n");
+
+    r = _bf_get_ctr_vals(chain_list, counters);
+    if (r < 0)
+        return bf_err_r(r, "Could not get ctr vals\n");
+
+    size_t marsh_size = num_counters * sizeof(struct bf_counter);
+
+    r = bf_marsh_new(counter_marsh, counters, marsh_size);
+    if (r < 0)
+        if (r < 0)
+            return bf_err_r(r, "Failed to make new marsh\n");
+
+    free(counters);
```

Comment:
Let the framework do that for you: create a new list to contain `bf_counter` structures. For each chain you allocate a counter, fill it with the data, push it to the list. Finally, `bf_list_marsh()` will serialize it for you.